src/extensions/compression/deflate.rs

```diff
@@ -0,0 +1,442 @@
+        for (key, val) in agreed {
+            match key {
+                SERVER_NO_CONTEXT_TAKEOVER => {
+                    // Fail the connection when the response contains multiple parameters with the same name.
+                    if seen_server_no_context_takeover {
+                        return Err(DeflateError::Negotiation(
+                            NegotiationError::DuplicateParameter(key.to_owned()),
+                        ));
+                    }
+                    seen_server_no_context_takeover = true;
+                    // A server MAY include the "server_no_context_takeover" extension
+                    // parameter in an extension negotiation response even if the extension
+                    // negotiation offer being accepted by the extension negotiation
+                    // response didn't include the "server_no_context_takeover" extension
+                    // parameter.
+                    config.server_no_context_takeover = true;
+                }
+
+                CLIENT_NO_CONTEXT_TAKEOVER => {
+                    // Fail the connection when the response contains multiple parameters with the same name.
+                    if seen_client_no_context_takeover {
+                        return Err(DeflateError::Negotiation(
+                            NegotiationError::DuplicateParameter(key.to_owned()),
+                        ));
+                    }
+                    seen_client_no_context_takeover = true;
+                    // The server may include this parameter in the response and the client MUST support it.
+                    config.client_no_context_takeover = true;
+                }
+
+                SERVER_MAX_WINDOW_BITS => {
+                    // Fail the connection when the response contains a parameter with invalid value.
+                    if let Some(bits) = val {
+                        if !is_valid_max_window_bits(bits) {
+                            return Err(DeflateError::Negotiation(
+                                NegotiationError::InvalidServerMaxWindowBitsValue(bits.to_owned()),
+                            ));
+                        }
+                    } else {
+                        return Err(DeflateError::Negotiation(
+                            NegotiationError::MissingServerMaxWindowBitsValue,
+                        ));
+                    }
+
+                    // A server may include the "server_max_window_bits" extension parameter
+                    // in an extension negotiation response even if the extension
+                    // negotiation offer being accepted by the response didn't include the
+                    // "server_max_window_bits" extension parameter.
+                    //
+                    // However, but we need to fail the connection because we don't support it (condition 4).
+                    return Err(DeflateError::Negotiation(
```

Comment:
This is incorrect. Server may include "server_max_window_bits" without the client asking for it for the explicit reason that it informs the window size for compression that the server uses. There is no harm for the client to ignore that value, and use a full sized window for decompression - and since our window is 15 bits by default now that it isn't configurable, any valid value from server is fine for the client.

The correct behaviour is to have the normal duplicate checking for this response header, but accept any valid value (or any value below the window size we use, but flate2 doesn't expose that so we will just need to assume it's 15 bits).

Reply:
Thanks for your comment.